internal/scheduler/scheduling_algo.go

```diff
@@ -190,42 +189,67 @@ func (l *LegacySchedulingAlgo) scheduleOnExecutor(
 		l.config,
 		totalCapacity,
 	)
-	queues := make([]*Queue, 0, len(priorityFactorByQueue))
-	for name, priorityFactor := range priorityFactorByQueue {
-		it := db.QueuedJobs(txn, name)
-		if queue, err := NewQueue(name, priorityFactor, &JobQueueIteratorAdapter{it: it}); err != nil {
-			return nil, err
-		} else {
-			queues = append(queues, queue)
-		}
-	}
-	legacyScheduler, err := NewLegacyScheduler(
-		ctx,
+	scheduler := NewRescheduler(
 		*constraints,
+		l.config,
+		&schedulerJobRepositoryAdapter{
+			txn: txn,
+			db:  db,
+		},
 		nodeDb,
-		queues,
-		totalResourceUsageByQueue,
+		priorityFactorByQueue,
+		allocatedByQueueAndPriority,
+		nil, // TODO: Add a repo to enable querying for scheduler reports.
 	)
+	result, err := scheduler.Schedule(ctx)
 	if err != nil {
 		return nil, err
 	}
-	result, err := legacyScheduler.Schedule(ctx)
-	if err != nil {
-		return nil, err
+
+	for i, job := range result.PreemptedJobs {
+		jobDbJob := job.(*jobdb.Job)
+		node := result.NodeByJobId[jobDbJob.GetId()]
+		if node == nil {
+			return nil, errors.Errorf("job %s not mapped to any node", jobDbJob.GetId())
+		}
+		if run := jobDbJob.LatestRun(); run != nil {
+			jobDbJob = jobDbJob.WithUpdatedRun(run.WithFailed(true))
+		}
+		result.ScheduledJobs[i] = jobDbJob.WithQueued(false).WithFailed(true)
 	}
-	updatedJobs := make([]*jobdb.Job, len(result.ScheduledJobs))
-	for i, report := range legacyScheduler.SchedulingRoundReport.SuccessfulJobSchedulingReports() {
-		job := report.Job.(*jobdb.Job)
-		nodeName := ""
-		if len(report.PodSchedulingReports) > 0 {
-			nodeName = report.PodSchedulingReports[0].Node.Name
-		} else {
-			log.Warnf("Could not resolve node for Job %s as no PodSchedulingReports were present", job.Id())
+	for i, job := range result.ScheduledJobs {
+		jobDbJob := job.(*jobdb.Job)
+		node := result.NodeByJobId[jobDbJob.GetId()]
+		if node == nil {
+			return nil, errors.Errorf("job %s not mapped to any node", jobDbJob.GetId())
 		}
-		job = job.WithQueued(false).WithNewRun(executor.Id, nodeName)
-		updatedJobs[i] = job
+		result.ScheduledJobs[i] = jobDbJob.WithQueued(false).WithNewRun(executor.Id, node.Name)
 	}
-	return updatedJobs, nil
+	return result, nil
+}
+
+type schedulerJobRepositoryAdapter struct {
+	db  *jobdb.JobDb
+	txn *jobdb.Txn
+}
+
+func (repo *schedulerJobRepositoryAdapter) GetQueueJobIds(queue string) ([]string, error) {
```

Comment:
This f

Reply:
If we do need these functions I'd suggest writing a bit of doc and renaming this one  to `GetQueuedJobIds` as it does fetches all queued jobs

Reply:
We need it to implement the `JobRepository interface`. I suggest leaving these for now and separately improving `Rescheduler` to take an iterator over queued job ids.